### PR TITLE
feat: init web3-marketplace-plugin

### PR DESCRIPTION
**Pull request by luu-alex**
https://www.notion.so/chainsafe/Web3js-Marketplace-Plugin-e52134db8223409b9003323924cd1b3a
this PR is for milestone 1. Implements marketplace contracts

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,41 +1,40 @@
...
-    "ts-jest": "^29.1.0",
-    "ts-node": "^10.9.1",
-    "typescript": "^5.1.3",
-    "web3": "^4.2.2"
-  },
-  "peerDependencies": {
-    "web3": ">= 4.0.3"
-  }
+	"name": "web3-marketplace-plugin",
+	"version": "1.0.0",
```
Comment (jdevcs) [inline package.json]:
```suggestion
	"version": "0.1.0",
```
dev version

File: package.json [hunk 2]
```diff
@@ -1,41 +1,40 @@
...
+	"description": "web3.js plugin for Chainsafe gaming marketplace",
+	"main": "lib/index.js",
+	"types": "lib/index.d.ts",
+	"homepage": "https://github.com/web3/web3.js-plugin-template#readme",
```
Comment (jdevcs) [inline package.json]: this also needs update with current repo

Comment (jdevcs) [reply package.json]: including other places in this file

File: src/marketplaceContract.ts [hunk 3]
```diff
@@ -0,0 +1,1362 @@
+export const marketplaceABI = [
```
Comment (jdevcs) [inline src/marketplaceContract.ts]: I'll suggest to also add solidity contract perm code link in comments

File: src/index.ts [hunk 4]
```diff
@@ -1,16 +1,177 @@
...
+			return receipt;
+		} catch (error) {
+			console.error('Error listing item:', error);
+			throw error;
+		}
+	}
+	/**
+	 * Places multiple items for sale on the marketplace using Web3.js.
+	 *
+	 * @param {Object} contract - The initialized Web3 contract instance of the marketplace.
```
Comment (danforbes) [inline src/index.ts]: Should this be a Contract?

```suggestion
	 * @param {Contract<typeof marketplaceABI>} contract - The initialized Web3 contract instance of the marketplace.
```

File: src/index.ts [hunk 5]
```diff
@@ -1,16 +1,177 @@
...
+		} catch (error) {
+			console.error('Error listing items:', error);
+			throw error;
+		}
+	}
+
+	/**
+	 * Cancels the listing of the token and returns it to the seller using Web3.js.
+	 *
+	 * @param {Contract<typeof marketplaceABI>} contract - The initialized Web3 contract instance of the marketplace.
```
Comment (danforbes) [inline src/index.ts]: Can the marketplace contract be pulled out into its own type? That may be nice for this plugin: `type MarketplaceContract = Contract<typeof marketplaceABI>`.

File: src/index.ts [hunk 6]
```diff
@@ -1,16 +1,177 @@
...
+				"No account found, please connect to a wallet provider or set a default account - 'web3.config.defaultAccount = <your address>'",
+			);
+		}
+		return sender;
+	};
+
+	/**
+	 * Places an item for sale on the marketplace using Web3.js.
+	 *
+	 * @param {Contract} contract - The initialized Web3 contract instance of the marketplace.
```
Comment (danforbes) [inline src/index.ts]:
```suggestion
	 * @param {Contract<typeof marketplaceABI>} contract - The initialized Web3 contract instance of the marketplace.
```

**Comment by luu-alex** [comment]
@jdevcs  thanks, updated

File: .prettierrc.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
+	"printWidth": 100,
+	"singleQuote": true,
+	"trailingComma": "all",
+	"useTabs": true,
+	"arrowParens": "avoid"
+}
```
Comment (danforbes) [inline .prettierrc.json:7]: Where did this come from? Do we use this same configuration in other projects? Is it compatible with the standard ChainSafe ESLint configuration?

File: src/index.ts [hunk 8]
```diff
@@ -1,16 +1,177 @@
...
+	async cancelListing(contract: Contract<typeof marketplaceABI>, itemId: string): Promise<Receipt> {
+		try {
+			const sender = await this.getSender();
+
+			// Send the transaction to call the `cancelListing` function
+			const receipt = await contract.methods.cancelListing(itemId).send({
+				from: sender,
+			});
+
+			console.log('Listing canceled successfully:', receipt);
```
Comment (danforbes) [inline src/index.ts]: Is this supposed to be here? I don't see `console.log` messages for any of the above functions.

File: .prettierrc.json (hunk 7 above)
Comment (luu-alex) [reply .prettierrc.json:7]: Its the same prettier config thats being used in web3.js 
https://github.com/web3/web3.js/blob/4.x/templates/.prettierrc.json.tmpl

File: src/index.ts (hunk 8 above)
Comment (luu-alex) [reply src/index.ts]: good catch, removing those

File: README.md [hunk 9]
```diff
@@ -1,34 +1,16 @@
-web3-plugin-template
-===========
+# web3-marketplace-plugin for Web3.js
 
-This is a template for creating a repository for web3.js plugin.
+The Marketplace plugin is for the Chainsafe gaming marketplace. This pliugin will enable web developers to utilize Web3.js with the NFT Launchpad and NFT marketplace contracts on the [gaming dashboard](https://dashboard.gaming.chainsafe.io/dashboard) in their web applications.
```
Comment (jdevcs) [inline README.md]:
```suggestion
The Marketplace plugin is for the Chainsafe gaming marketplace. This plugin will enable web developers to utilize Web3.js with the NFT Launchpad and NFT marketplace contracts on the [gaming dashboard](https://dashboard.gaming.chainsafe.io/dashboard) in their web applications.
```
typo

File: src/marketplacePlugin.ts [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+		try {
+			const sender = await this.getSender();
+
+			const receipt = await contract.methods.listItem(nftContract, tokenId, price, deadline).send({
+				from: sender,
+			});
+
+			return receipt;
+		} catch (error) {
+			console.error('Error listing item:', error);
```
Comment (jdevcs) [inline src/marketplacePlugin.ts]: I'll suggest to remove all consol.error

**Review by jdevcs — APPROVED** [review]
suggested few minor changes, PR lgtm

**Pull request merged**